### PR TITLE
Fixes for intra-process actions

### PR DESCRIPTION
**Pull request by mauropasse**
Fixes for intra-process actions:
- Don't throw in a normal situation of expired IPC Action client
- Check if data in buffers before extract
- Fix IPC Actions QoS depth
- Fix data race on IPC Actions callbacks
- Correct intra-process actions is_ready()
- Use Goal ID as key to store client callbacks and responses from server
- Use the FNV-1a hash algorithm for Goal UUID
- Remove RCL_RET_CLIENT_TAKE_FAILED error log - add comment instead

The following (simplified) flowchart represents what's happening when the Action Client sends a goal request go the Action Server, until the server accepts and responds to the client:
```
auto goal_handle_future = action_client->async_send_goal(goal_msg, send_goal_options);
auto goal_handle = goal_handle_future.get();
```

image

The process is almost exactly the same for:
```
auto result_future = action_client->async_get_result(goal_handle);
auto wrapped_result = result_future.get();
```
and for cancel:
```
    auto cancel_result_future = action_client->async_cancel_goal(goal_handle);
    auto cancel_result = cancel_result_future.get();
```
In the following chart I show part of `action_client->async_get_result` but focusing on the server logic, which sends the result only if the client has requested for it:
image

### CODE REVIEW

File: rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp [hunk 1]
```diff
@@ -100,93 +118,110 @@ class ActionClientIntraProcess : public ActionClientIntraProcessBase
...
-    is_cancel_response_ready_ = true;
-    invoke_on_ready_callback(EventType::CancelResponse);
+    invoke_on_ready_callback(EventType::CancelResponse, goal_id);
   }
 
   void store_ipc_action_feedback(FeedbackSharedPtr feedback)
   {
     feedback_buffer_->add(std::move(feedback));
     gc_.trigger();
-    is_feedback_ready_ = true;
```
Comment (alsora) [inline rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp:157]: why these are removed?

File: rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp [hunk 2]
```diff
@@ -195,79 +230,143 @@ class ActionClientIntraProcess : public ActionClientIntraProcessBase
...
+    ResultResponsePairSharedPtr ready_response;
+
+    for (auto & response : responses) {
+      auto goal_id = response->first;
+      // Get the first response which "is ready", meaning that we have
+      // the response callback to process the event
+      if (!ready_response && goal_has_response_callback(goal_id)) {
+        ready_response = std::move(response);
+      } else {
+        // Not ready, re-add to the buffer
```
Comment (alsora) [inline rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp]: I have a few concerns about this code.
 - could it be more efficient to operate on a single element?
 - is this thread safe?
 - what happens if "something else" becomes ready while we are doing this work

**Comment by mauropasse** [comment]
Some comments about how things work, I'll try to make it simple and not very long. (see new flowchart in comment below)

Every time the `Client` requests something to the `Server`, a callback is created: it will be called when the server sends back the response (using the response as argument).
After the callback is called, must be removed (in some cases), since they maintain in scope shared pointers.

We have [5 types of Action Client events](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp_action/include/rclcpp_action/client.hpp#L75-L82) (`EventType`), all of them with their `ResponseCallback` callback:
* `FeedbackSubscription` -> [Callback](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp#L84) created on client constructor - no need to remove after call
* `StatusSubscription`  -> [Callback](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp#L83) created on client constructor - no need to remove after call.

For every individual goal sent we have: 
* `GoalClient` -> [Callback](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp_action/include/rclcpp_action/client.hpp#L471) created when `async_send_goal` - must be removed after call
* `ResultClient` ->  [Callback](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp_action/include/rclcpp_action/client.hpp#L809) created when `async_get_result` - must be removed after call
* `CancelClient`  -> [Callback](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp_action/include/rclcpp_action/client.hpp#L877) created when `async_cancel_goal`  - must be removed after call


Since a client can make multiple requests, we need a storage for all the individual goal IDs, event types, their callbacks, and the `unread_count`.

I created [a structure](https://github.com/mauropasse/rclcpp/blob/ecf8f955eb469741db9762ad26e5512493352dbb/rclcpp/include/rclcpp/experimental/action_client_intra_process_base.hpp#L185) to hold all the info and data to process the different events, mapped with their respective "Goal ID". So we have in the map:

` < Goal_ID, {EventType, ResponseCallback, unread_count} >`

Besides this map, we have all the [IPC ring-buffers](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp#L355-L369) to hold the responses from the server. They look like:

` <Goal_ID, ServerResponse> `

So when we extract an element (response) form the ring buffer, we get the `Goal_ID` and use it as key to extract the `ResponseCallback`  and call it (and then remove it, if necessary).

**Comment by bpwilcox** [comment]
> For a simple interaction client/server, there are usually 4 threads involved. That's an extra complication that required attention avoiding races

Could you elaborate which threads are involved and what each are doing for a client/server interaction?

**Comment by mauropasse** [comment]
> Could you elaborate which threads are involved and what each are doing for a client/server interaction?

The threads involved are:
1. Client app thread sending goal requests
2. Client executor thread (sets the "on_ready_callback" and executes action client work)
3. Server executor thread (puts responses into the client's ring buffers)
4. Server thread (usually created to execute the goal, also puts responses into the client's ring buffers)

File: rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp (hunk 1 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp:157]: Setting the flag here was breaking the `SingleThreadedExecutor`, which already sets the `is_*_ready_` flags on the [is_ready()](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp#L117-L132) API.
For `EventsExecutor`, the flags are set on the [take_data_by_entity_id()](https://github.com/mauropasse/rclcpp/blob/mauro/fix-ipc-actions-squashed/rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp#L227-L250) API.

File: rclcpp_action/include/rclcpp_action/client.hpp [hunk 3]
```diff
@@ -527,29 +531,37 @@ class Client : public ClientBase
...
+    std::lock_guard<std::mutex> guard(goal_handles_mutex_);
+    auto goal_handle_it = goal_handles_.begin();
+    while (goal_handle_it != goal_handles_.end()) {
+      if (!goal_handle_it->second.lock()) {
+        RCLCPP_DEBUG(
+          this->get_logger(),
+          "Dropping weak reference to goal handle during send_goal()");
+        goal_handle_it = goal_handles_.erase(goal_handle_it);
+
+        size_t hashed_guuid = std::hash<GoalUUID>()(goal_handle_it->first);
```
Comment (bpwilcox) [inline rclcpp_action/include/rclcpp_action/client.hpp]: Is this the correct order, (e.g. to get hash of _next_  goal handle _after_ the previous one was erased)? You then attempt to erase the goal info from the ipc client.

Comment (mauropasse) [reply rclcpp_action/include/rclcpp_action/client.hpp]: Yes wrong order, well spotted! Fixed in https://github.com/irobot-ros/rclcpp/pull/144/commits/62e694b87ca6ed2cf43a7168ba5b02ba99b83d29

File: rclcpp/include/rclcpp/experimental/intra_process_manager.hpp [hunk 4]
```diff
@@ -549,30 +553,23 @@ class IntraProcessManager
...
-    // Now lets send the goal request
     auto service = get_matching_intra_process_action_server<ActionT>(ipc_action_client_id);
 
     if (service) {
       service->store_ipc_action_goal_request(
         ipc_action_client_id, std::move(goal_request));
     }
+
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+    clients_uuid_to_id_[goal_id] = ipc_action_client_id;
```
Comment (bpwilcox) [inline rclcpp/include/rclcpp/experimental/intra_process_manager.hpp]: If the service is `nullptr`, is it still valid to store the goal id here?

File: rclcpp_action/include/rclcpp_action/server.hpp [hunk 5]
```diff
@@ -483,20 +489,11 @@ class Server : public ServerBase, public std::enable_shared_from_this<Server<Act
...
-    }
+    auto ipm = lock_intra_process_manager();
     ipm->remove_action_server(ipc_action_server_id_);
   }
 
 protected:
-  // Intra-process version of execute_goal_request_received_
-  // Missing: Deep comparison of functionality betwen IPC on/off
   void
   ipc_execute_goal_request_received(GoalRequestDataPairSharedPtr data)
```
Comment (bpwilcox) [inline rclcpp_action/include/rclcpp_action/server.hpp:498]: In this ipc function I still see calls to `rcl` such as `rcl_action_get_zero_initialized_goal_info()`. What is the `goal_info`? How is this relevant or necessary to call into rcl when going through ipc?

File: rclcpp_action/include/rclcpp_action/server.hpp [hunk 6]
```diff
@@ -564,21 +549,13 @@ class Server : public ServerBase, public std::enable_shared_from_this<Server<Act
...
-    auto ipm = weak_ipm_.lock();
-    if (!ipm) {
-      throw std::runtime_error(
-              "intra_process_action_send_cancel_response called "
-              " after destruction of intra process manager");
-    }
+    auto ipm = lock_intra_process_manager();
 
     // Convert c++ message to C message
     rcl_action_cancel_request_t cancel_request = rcl_action_get_zero_initialized_cancel_request();
```
Comment (bpwilcox) [inline rclcpp_action/include/rclcpp_action/server.hpp:561]: Similar question to my other comment, but why do we need to get `cancel_request` from `rcl` layer while doing ipc? I know this PR is built on top of previous work but I am missing the rationale for the interaction with the rcl layer.

File: rclcpp/include/rclcpp/experimental/intra_process_manager.hpp (hunk 4 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/experimental/intra_process_manager.hpp]: No, the `goal_id` should be stored only if `service` is valid.
Now that I see, if `service` expired, I should return a `false` to let the caller know that it couldn't send intra-process, and should proceed with inter-process.
Both are fixed in https://github.com/irobot-ros/rclcpp/pull/144/commits/ecf8f955eb469741db9762ad26e5512493352dbb.

File: rclcpp_action/include/rclcpp_action/server.hpp (hunk 6 above)
Comment (mauropasse) [reply rclcpp_action/include/rclcpp_action/server.hpp:561]: Only the communication (send request / responses / etc) goes through intra-process.
But the rest of the logic still lives in `rcl`, that is, we still use the `rcl_handle` which controls the goal state, etc.

In summary, all the `rcl_action_send_*` have their parallel `intra_process_action_send_*` versions. But the rest of the code is common.

File: rclcpp_action/include/rclcpp_action/server.hpp (hunk 5 above)
Comment (mauropasse) [reply rclcpp_action/include/rclcpp_action/server.hpp:498]: In this case, the `rcl_action_goal_info_t goal_info` is just used to obtain the `rcl_goal_handle`, to then update the goal state. 
The "bookkeeping" of the goal state is still performed in `rcl`.

File: rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp (hunk 2 above)
Comment (mauropasse) [reply rclcpp/include/rclcpp/experimental/action_client_intra_process.hpp]: I've removed this function, now the server holds the response until the client requests it, avoiding the need for this.

**Comment by mauropasse** [comment]
The following flowchart represents what's happening when the Action Client sends a goal request go the Action Server, until the server accepts and responds to the client:
```
auto goal_handle_future = action_client->async_send_goal(goal_msg, send_goal_options);
auto goal_handle = goal_handle_future.get();
```

image

The process is almost exactly the same for:
```
auto result_future = action_client->async_get_result(goal_handle);
auto wrapped_result = result_future.get();
```
and for cancel:
```
    auto cancel_result_future = action_client->async_cancel_goal(goal_handle);
    auto cancel_result = cancel_result_future.get();
```

**Review by alsora — APPROVED** [review]
